eli5/formatters/html.py

```diff
@@ -49,19 +51,29 @@ def format_html_styles():
     return template_env.get_template('styles.html').render()
 
 
-def render_weighted_spans(weighted_spans_data):
+def render_weighted_spans(weighted_spans_data, preserve_density=None):
     """ Render text document with highlighted features.
+    If preserve_density is True, then color for longer fragments will be
+    less intensive than for shorter fragments, so that "sum" of intensities
+    will correspond to feature weight.
+    If preserve_density is None, then it's value is chosen depending on
+    analyzer kind: it is preserved for "char" and "char_wb" analyzers,
+    and not preserved for "word" analyzers.
     """
+    if preserve_density is None:
+        preserve_density = weighted_spans_data['analyzer'].startswith('char')
     doc = weighted_spans_data['document']
     weighted_spans = weighted_spans_data['weighted_spans']
     char_weights = np.zeros(len(doc))
     for _, spans, weight in weighted_spans:
         for start, end in spans:
+            if preserve_density:
+                weight /= (end - start)
             char_weights[start:end] += weight
     # TODO - can be much smarter, join spans at least
     # TODO - for longer documents, remove text without active features
     not_found_weights = sorted(
-        (feature, weight)
+        (feature, weight / len(feature) if preserve_density else weight)
```

Comment:
This part is tricky - non-text feature names can be lighter or darker depending on a length of their name; two features with equal weights can have different color if their names have different length. I don't know how to solve it :) My main motivation was to make highlighting compatible between char and token models, not so much between text and non-text features.

Maybe it deserves its own solution - e.g. instead of showing non-text features inline before the text we may show them separated by new lines, and near each feature name show a rectangle, all rectangles of the same size. This way length of non-text feature name won't impact highlighting. To make it compatible with text rectangle may have a width similar to avg. word width (e.g. 5 chars).